### PR TITLE
Draft: Support out of order byte-math keywords

### PR DESCRIPTION
**Pull request by jlucovsky**
This is a draft PR for the Rust based byte-math keyword parser that supports arbitrary order keyword values.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [5077](https://redmine.openinfosecfoundation.org/issues/5077)

Describe changes:
- Rust based parser for byte-math keywords with tests
- Removal of the Regex-based byte-math C parser
- Retention of the C-based unittests from the previous implementation (for catching regressions of the Rust parser replacement).

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: rust/src/detect_parser/byte_math.rs [hunk 1]
```diff
@@ -0,0 +1,1043 @@
...
+    numeric_val: u32,
+}
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct DetectByteMathData {
+    pub local_id: u8,
+    pub nbytes: u8,
+    pub offset: i32,
+    pub oper: u8,
```
Comment (victorjulien) [inline rust/src/detect_parser/byte_math.rs:127]: nit: reorganize for better alignment and avoiding padding

File: rust/src/detect_parser/byte_math.rs [hunk 2]
```diff
@@ -0,0 +1,1043 @@
...
+            ResultValue {
+                numeric: true,
+                string_val: "".to_string(),
+                numeric_val: val as u32,
+            },
+        ))
+    } else {
+        Ok((
+            input,
+            ResultValue {
```
Comment (victorjulien) [inline rust/src/detect_parser/byte_math.rs:267]: I think ResultValue can be an enum.

File: rust/src/detect_parser/byte_math.rs [hunk 3]
```diff
@@ -0,0 +1,1043 @@
...
+pub const DETECT_BYTEMATH_BASE_DEC: u8 = 10;
+pub const DETECT_BYTEMATH_BASE_OCT: u8 = 8;
+pub const DETECT_BYTEMATH_BASE_NONE: u8 = 0;
+pub const DETECT_BYTEMATH_BASE_DEFAULT: u8 = DETECT_BYTEMATH_BASE_DEC;
+
+/// Custom rule parse errors.
+///
+/// Implemented based on the Nom example for implementing custom errors.
+#[derive(Debug, PartialEq)]
+pub enum RuleParseError<I> {
```
Comment (victorjulien) [inline rust/src/detect_parser/byte_math.rs:61]: lets only add things that are actually used here

File: rust/src/detect_parser/byte_math.rs [hunk 4]
```diff
@@ -0,0 +1,1043 @@
...
+}
+// Fixed position parameter count: bytes, offset, oper, rvalue, result
+// result is not parsed with the fixed position parameters as it's
+// often swapped with optional parameters
+pub const DETECT_BYTEMATH_FIXED_PARAM_COUNT: u8 = 5;
+// Optional parameters: endian, relative, string, dce, bitmask
+pub const DETECT_BYTEMATH_MAX_PARAM_COUNT: u8 = 10;
+
+#[derive(Debug)]
+struct ResultValue {
```
Comment (victorjulien) [inline rust/src/detect_parser/byte_math.rs:115]: This should probably be something like
```rust
enum ResultValue {
   NumericVal(u32);
   StringVal(String);
}
```
Then you don't need the `bool` and Rust can take care of the type checking.

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpw1_files_sha256.

ERROR: QA failed on tlpr1_alerts_cmp.
  

Pipeline 7699

File: rust/src/detect_parser/byte_math.rs (hunk 3 above)
Comment (jlucovsky) [reply rust/src/detect_parser/byte_math.rs:61]: I'm moving these into a separate module and will trim accordingly.

File: rust/src/detect_parser/byte_math.rs [hunk 5]
```diff
@@ -0,0 +1,1010 @@
...
+// often swapped with optional parameters
+pub const DETECT_BYTEMATH_FIXED_PARAM_COUNT: u8 = 5;
+// Optional parameters: endian, relative, string, dce, bitmask
+pub const DETECT_BYTEMATH_MAX_PARAM_COUNT: u8 = 10;
+
+#[derive(Debug)]
+enum ResultValue {
+    NumericVal(u64),
+    StringVal(String),
+}
```
Comment (jasonish) [inline rust/src/detect_parser/byte_math.rs:70]: There's a bit of a stutter here.  Could juse use `Numeric` and `String`.

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpw1_files_sha256.

ERROR: QA failed on tlpr1_alerts_cmp.

ERROR: QA failed on ips_afp_drop_chk.
  

Pipeline 7706

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpw1_files_sha256.

ERROR: QA failed on tlpr1_alerts_cmp.
  

Pipeline 7730

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpr1_alerts_cmp.
  

Pipeline 7738

**Comment by jlucovsky** [comment]
Continued in #7509

**Pull request closed without merging**